cleverhans/experimental/certification/nn.py

```diff
@@ -43,22 +47,70 @@ def __init__(self, net_weights, net_biases, net_layer_types):
     self.biases = []
     self.layer_types = []
     self.sizes = []
+    self.input_shapes = []
+    self.output_shapes = []
+    self.has_conv = False
+    if(input_shape is not None):
+      current_num_rows = input_shape[0]
+      current_num_columns = input_shape[1]
+      current_num_channels = input_shape[2]
+
     # Setting the sizes of the layers of the network
     # sizes[i] contains the size of x_i
     for i in range(self.num_hidden_layers):
       shape = np.shape(net_weights[i])
-      self.sizes.append(int(shape[1]))
+      # self.sizes.append(int(shape[1]))
       self.weights.append(
           tf.convert_to_tensor(net_weights[i], dtype=tf.float32))
-      self.biases.append(tf.convert_to_tensor(net_biases[i], dtype=tf.float32))
+      # self.biases.append(tf.convert_to_tensor(net_biases[i], dtype=tf.float32))
       self.layer_types.append(net_layer_types[i])
 
-    # Last layer shape
-    self.sizes.append(int(np.shape(net_weights[self.num_hidden_layers - 1])[0]))
-    self.final_weights = tf.convert_to_tensor(
-        net_weights[self.num_hidden_layers], dtype=tf.float32)
-    self.final_bias = tf.convert_to_tensor(
-        net_biases[self.num_hidden_layers], dtype=tf.float32)
+      if(self.layer_types[i] in {'ff', 'ff_relu'}):
+        self.sizes.append(int(shape[1]))
+        # For feedforward networks, no unraveling the bias terms
+
+        small_bias = tf.convert_to_tensor(net_biases[i], dtype=tf.float32)
+        self.biases.append(tf.reshape(small_bias, [-1, 1]))
+        # Assumes that x^{i+1} = W_i x^i
+        self.input_shapes.append([int(shape[1]), 1])
+        self.output_shapes.append([int(shape[0]), 1])
+
+      # Convolution type
+      else:
+        self.has_conv = True
+        num_filters = shape[3]
+        self.input_shapes.append([1, current_num_rows, current_num_columns, current_num_channels])
+        self.sizes.append(current_num_rows*current_num_columns*current_num_channels)
+        current_num_channels = num_filters
+        # For propagating across multiple conv layers
+        if(PADDING == 'SAME'):
+          current_num_rows = int(current_num_rows/STRIDE)
+          current_num_columns = int(current_num_columns/STRIDE)
+        self.output_shapes.append(
+            [1, current_num_rows, current_num_columns, current_num_channels])
+
+        # For conv networks, unraveling the bias terms
+        small_bias = tf.convert_to_tensor(net_biases[i], dtype=tf.float32)
+        large_bias = tf.tile(tf.reshape(small_bias, [-1, 1]),
+                             [current_num_rows*current_num_columns, 1])
+        self.biases.append(large_bias)
+
+    # Last layer shape: always ff
+    if self.has_conv:
```

Comment:
Most of the code in this if statement is duplicated, except for computation of size.
So you can refactor it in a following way:

```
if self.has_conv:
  final_dim = ...
else:
  final_dim = ...
self.sizes.append(final_dim)
...
```